opt: normalize sequential % in regexes to single %

sriram2000na opened the pull request:
Cloned #80432, and called the created function inside scalar_funcs.go. Have added tests in testdata/rules/scalar. I don't know 
how the test output is supposed to be written. Please help.
Epic: None
Issue:https://github.com/cockroachdb/cockroach/issues/80192

ADustyOldMuffin committed:
opt: normalize sequential '%' in regexes

Fixes: #80192
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124263)
<!-- Reviewable:end -->

mgartner commented on pkg/sql/opt/norm/rules/scalar.opt:
```diff
@@ -441,3 +441,12 @@ $value
 =>
 (FetchVal $input $index)
+
+
```
nit: remove extra line

mgartner commented on pkg/sql/opt/norm/testdata/rules/scalar:
```diff
@@ -2441,3 +2441,15 @@ project
       └── b.arr:5[2] [as=arr:9, outer=(5)]
+
+norm expect=NormalizeLikePattern
+SELECT i FROM a WHERE s LIKE '%%abc%%%%dddd'
```
We should have additional tests for `NOT LIKE`, `ILIKE`, `NOT ILIKE`, `SIMILAR TO`, and `NOT SIMILAR TO`, and a test where the rule does not match.

nit: I'd also simplify this to `SELECT s FROM a WHERE s LIKE ...` or `SELECT s LIKE '...' FROM a`

mgartner commented on pkg/sql/opt/norm/rules/scalar.opt:
```diff
@@ -441,3 +441,12 @@ $value
 (FetchVal $input $index)
+
+
+[NormalizeLikePattern, Normalize]
```
I think this rule (and the corresponding tests and golang functions) belongs in comp.opt instead.

mgartner commented on pkg/util/strings.go:
```diff
@@ -167,3 +167,37 @@ func RandString(rng *rand.Rand, length int, alphabet string) string {
+// and collapse any repeating instances of this character to a
+// single instance.
+// E.g. CollapseDupeChar("wwwhy hello there", 'w') returns "why hello there"
+func CollapseDupeChar(toCollapse string, collapseChar rune) string {
```
I also think `CollapseRepeatedChar` might be a better name to signify that it's specifically collapsing adjacent `collapseChar`s.

sriram2000na replied:
Done

mgartner commented on pkg/util/strings_test.go:
```diff
@@ -141,3 +145,78 @@ func TestStringListBuilder(t *testing.T) {
+		{"%test%%%%", '%', "%test%"},
+		{"%%%test%%%%", '%', "%test%"},
+		{"%%%test1%%%test2%%%test3%%%", '%', "%test1%test2%test3%"},
+		{"I work on ddddddifferent characters", 'd', "I work on different characters"},
```
Can you add tests with multi-byte characters?

mgartner commented on pkg/sql/opt/norm/scalar_funcs.go:
```diff
@@ -384,3 +386,14 @@ func (c *CustomFuncs) SplitTupleEq(lhs, rhs *memo.TupleExpr) memo.FiltersExpr {
 }
+
+func (c *CustomFuncs) CollapseDupeChar(input opt.ScalarExpr) opt.ScalarExpr {
+	pattern := memo.ExtractConstDatum(input).String()
```
We don't want to use the `String()` method here, since this formats the `DString` which is costly and unnecessary. Instead, we should use a type assertion:

```go
	s, ok := memo.ExtractConstDatum(cnst).(*tree.DString)
	if !ok {
		panic(errors.AssertionFailedf("expected string constant")
	}
```

mgartner reviewed (commented):
Thanks for the contribution! I've left a few initial comments.

Also, please combine the commits into one and follow the commit message guidelines here: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#General-structure-of-a-commit-message

sriram2000na commented:
> Thanks for the contribution! I've left a few initial comments.
> 
> Also, please combine the commits into one and follow the commit message guidelines here: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#General-structure-of-a-commit-message

Thank you for reviewing, I am starting to work on the requested changes.

sriram2000na commented on pkg/util/strings_test.go:
```diff
@@ -141,3 +145,78 @@ func TestStringListBuilder(t *testing.T) {
+		runBenchmarkNoDupe,
+		runBenchmarkSingleDupe,
+		runBenchmarkMultipleDupe,
+		runBenchmarkSpacedDupe,
```
It made it harder to implement it without slice of functions. If this change is absolutely necessary, I can make it. Reflection didn't have the ability to get package level functions from the string name alone to run it.

mgartner replied:
I was thinking something like this:

```go
func BenchmarkCollapseRepeatedChar(b *testing.B) {
	type testCase struct {
		name  string
		input string
	}
	testCases := []testCase{
		{"no-dupe", "%test%"},
		{"single-dupe-group", "%test%%%%"},
		{"two-dupe-groups", "%%%%%test%%%%"},
		{"three-dupe-groups", "%%%spaced%%%dupe%%%"},
	}
	for _, tc := range testCases {
		b.Run(tc.name, func(b *testing.B) {
			for n := 0; n < b.N; n++ {
				CollapseRepeatedChar(tc.input, '%')
			}
		})
	}
}
```

sriram2000na replied:
Sorry for the late reply. Done. Please check.

mgartner committed:
opt: optimize norm rule for collapsing wildcards in LIKE pattern

The `NormalizeLikePattern` rule has been renamed and refatored. Also,
the `CollapseRepeatedRune` function (formerly named
`CollapseRepeatedChar`) has been optimized to avoid allocating new
strings when possible. The benchmark results below show the improvement.

```
name                                old time/op    new time/op    delta
CollapseRepeatedRune/no-target        62.6ns ± 1%    12.5ns ± 0%   -80.10%  (p=0.008 n=5+5)
CollapseRepeatedRune/no-dupe          89.1ns ± 9%    60.2ns ± 0%   -32.42%  (p=0.016 n=5+4)
CollapseRepeatedRune/start-dupe        102ns ± 0%      88ns ± 0%   -13.53%  (p=0.008 n=5+5)
CollapseRepeatedRune/end-dupe         96.6ns ± 0%    69.4ns ± 3%   -28.10%  (p=0.008 n=5+5)
CollapseRepeatedRune/multiple-dupe     212ns ± 0%     102ns ± 0%   -52.07%  (p=0.008 n=5+5)

name                                old alloc/op   new alloc/op   delta
CollapseRepeatedRune/no-target         2.00B ± 0%     0.00B       -100.00%  (p=0.008 n=5+5)
CollapseRepeatedRune/no-dupe           2.00B ± 0%     0.00B       -100.00%  (p=0.008 n=5+5)
CollapseRepeatedRune/start-dupe        2.00B ± 0%     0.00B       -100.00%  (p=0.008 n=5+5)
CollapseRepeatedRune/end-dupe          2.00B ± 0%     0.00B       -100.00%  (p=0.008 n=5+5)
CollapseRepeatedRune/multiple-dupe     50.0B ± 0%     48.0B ± 0%    -4.00%  (p=0.008 n=5+5)

name                                old allocs/op  new allocs/op  delta
CollapseRepeatedRune/no-target          1.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)
CollapseRepeatedRune/no-dupe            1.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)
CollapseRepeatedRune/start-dupe         1.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)
CollapseRepeatedRune/end-dupe           1.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)
CollapseRepeatedRune/multiple-dupe      2.00 ± 0%      1.00 ± 0%   -50.00%  (p=0.008 n=5+5)
```

Release note: None

mgartner commented on pkg/sql/opt/norm/testdata/rules/comp:
```diff
@@ -1087,3 +1087,101 @@ project
+ ├── scan a
+ │    └── columns: s:4
+ └── filters
+      └── s:4 LIKE e'\\%a\\%bc\\%' [outer=(4), constraints=(/4: [/'%a%bc%' - /'%a%bc%']; tight), fd=()-->(4)]
```
This transformation is invalid, see the TODO.

mgartner reviewed (changes requested):
Thanks for working on this! I refactored a few things and optimized the `CollapseRepeatedRune` function to avoid allocations when possible—see the newly added commit.

Unfortunately, after discussing with @michae2, we realized that the rule is currently invalid due to it not properly handling escaped wildcard characters, e.g., `\%`. See the inline comment.
